---
title: "Reference-based Label Prediction and Mapping of Porcine Ileal Cells: Creation of a Porcine PBMC Reference Dataset"
author: "Jayne Wiarda"
date: "5Apr2021"
output: 
  github_document: 
    keep_html: TRUE
---

Since there are already some pretty spectacular and well-annotated scRNA-seq datasets available in literature, we can utilize these external datasets as references we can compare to our porcine ileum scRNA-seq data, which we treat as a dataset to query. We can perform reference-based label transfer to predict the most similar cellular counterparts in already annotated datasets we use as references (Which reference cells are our query cells most similar to?), and we can further calculate mapping scores to see how well query data (our porcine ileum scRNA-seq data) is represented by the reference dataset (How well does the reference data truly resemble the query data?).

Unfortunately, we are the first to create a scRNA-seq dataset of cells in the porcine intestine, so we have to compare to other reference datasets either within the same species but across anatomical locations or within the same tissue but across species. Here are the three relevant datasets we found:

* A scRNA-seq dataset cataloging porcine PBMCs (Herrera-Uribe & Wiarda et al. 2021); referred to as SsPBMC_HerreraWiarda) 
  - Cells are from seven conventional pigs of various ages, including two pigs (ExpD) which are the same as the ones we collected ileum samples from for our ileum scRNA-seq dataset
  - Data can be found at locations specified in the manuscript, but we recovered the data in-house from our previous analyses
  - We removed one cluster labelled as 'Unknown' (cluster 34) prior to creating our reference
* A scRNA-seq dataset cataloging cells from human terminal ileum (Elmentaite et al. 2020; referred to as HsIleum_Elmentaite)
  - Cells are from 8 healthy controls and 7 samples from individuals with Crohn's disease
  - Data were downloaded as pediatric_RAWCOUNTS_cellxgene.h5ad from at https://www.gutcellatlas.org/#fetal
  - Prior to creating the reference, we selected only cells with an inferred state of control for creating our reference, thus excluding cells from Crohn's disease patients, and retained only one-to-one gene orthologs between pig and human
* A scRNA-seq dataset cataloging cells from murine ileum (Xu et al. 2019; referred to as MmIleum_Xu)
  - Cells are from healthy or allergy model mice
  - Data were downloaded from the Single Cell Portal, found at https://singlecell.broadinstitute.org/single_cell/study/SCP210/fasi-immune-mouse-small-intestine#study-summary
  - Prior to creating the reference, we selected only cells from control mice, thus exluding cells from mice modelling an allergic response, and retained only one-to-one gene orthologs between pig and mouse
  
In this section, we go over the steps involved in creating a reference dataset from porcine PBMCs.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

### Load required software packages

The R package, Seurat (Stuart & Butler et al. 2019; Hao et al. 2020), is utilized to perform label transfer and mapping. Note, this analysis is done with an upgraded version of Seurat compared to other analyses, and is also performed on another maching. We also utilize the R package, future (Bengtsson 2020), to parallelize some of the steps performed herein. Refer to sessionInfo() at the bottom of the page for the R and package versions used and additional information.

Additional packages used: ggplot2 (Wickham 2016)

```{r, message = FALSE}
library(ggplot2)
library(Seurat)
library(future)
```

## Create porcine PBMC reference dataset

### Load relevant data:
```{r, warning = FALSE, error = FALSE}
load('/Users/jayne.wiarda/Desktop/scRNAseq/scRNAseqIleumAtlas/Ileum/ReferenceData/HerreraWiarda_SsPBMC/PBMC7IntegratedNormalized_11042020.RData') # load in the R data saved in-house from previous manuscript work
```

### Allow parallelization 
```{r, warning = FALSE, error = FALSE}
plan("multiprocess", workers = 4) # change plan to allow parallelization with 4 workers
options(future.globals.maxSize = 5000 * 1024^2)
```

### Remove unneccesary information

```{r, warning = FALSE, error = FALSE}
SsPBMC_HerreraWiarda <- DietSeurat(All.integrated, 
                                   assays = 'RNA', # remove SCT and integrated assays
                                   dimreducs = NULL) # remove former dimensionality reductions
SsPBMC_HerreraWiarda # let's see what we have here...
```

### Remove cells with unknown identity

Remove cluster 34, which had an unknown identity in the porcine PBMC manuscript:

```{r, warning = FALSE, error = FALSE}
Idents(SsPBMC_HerreraWiarda) <- SsPBMC_HerreraWiarda$seurat_clusters # change identities to cluster numbers from publication
SsPBMC_HerreraWiarda <- subset(SsPBMC_HerreraWiarda, 
                     idents = c('34'),  # remove cells belonging to these clusters that were specified as unknown in publication
                     invert = TRUE) # exclude specified identities
SsPBMC_HerreraWiarda # now see we have fewer cells
```

### Re-normalize and re-integrate the data

Split the data into individual samples:

```{r, warning = FALSE, error = FALSE, message = FALSE}
SsPBMC_HerreraWiarda.list <- SplitObject(SsPBMC_HerreraWiarda, split.by = "SampleID") # split into the original samples that were processed for scRNA-seq
```

For each sample, normalize the data and find the top 2,000 most variable genes:

```{r, warning = FALSE, error = FALSE}
for (i in 1:length(SsPBMC_HerreraWiarda.list)) { # for each sample individually, let's normalize the data and find the 2000 most highly variable features
  SsPBMC_HerreraWiarda.list[[i]] <- NormalizeData(SsPBMC_HerreraWiarda.list[[i]], 
                                        verbose = TRUE, 
                                        normalization.method = "LogNormalize", 
                                        scale.factor = 10000, 
                                        assay = "RNA")
  SsPBMC_HerreraWiarda.list[[i]] <- FindVariableFeatures(SsPBMC_HerreraWiarda.list[[i]], 
                                               selection.method = "vst", 
                                               nfeatures = 2000, 
                                               verbose = TRUE)
}
```

Integrate the samples together:

```{r, warning = FALSE, error = FALSE, message = FALSE}
SsPBMC_HerreraWiarda.anchors <- FindIntegrationAnchors(object.list = SsPBMC_HerreraWiarda.list, 
                                             dims = 1:30) # find integration anchors between samples based on variable features for each sample with default parameters
referenceSsPBMC_HerreraWiarda <- IntegrateData(anchorset = SsPBMC_HerreraWiarda.anchors, 
                                     dims = 1:30) # integrate the data together based on integration anchors found with default parameters
```

### Re-scale data

```{r, warning = FALSE, error = FALSE, message = FALSE}
referenceSsPBMC_HerreraWiarda <- ScaleData(referenceSsPBMC_HerreraWiarda, 
                                 verbose = TRUE, 
                                 assay = 'integrated') # scale the genes in the integrated assay
```

### Re-run PCA analysis

Calculate the PCs of the integrated dataset:

```{r, warning = FALSE, error = FALSE, message = FALSE}
referenceSsPBMC_HerreraWiarda <- RunPCA(referenceSsPBMC_HerreraWiarda, # calculate first 100 PCs
                              npcs = 100, 
                              verbose = TRUE)
```

Visualize our PCs in an elbow plot:

```{r, warning = FALSE, error = FALSE}
ElbowPlot(referenceSsPBMC_HerreraWiarda,
          ndims = 100) # look at this plot to find the 'elbow' for significant PCs... use this number of PCs for creating UMAP, tSNE, & cell neighbors & clustering
```

Quantitiatively calculate how many PCs to use:

```{r, warning = FALSE, error = FALSE}
pct <- referenceSsPBMC_HerreraWiarda[["pca"]]@stdev / sum(referenceSsPBMC_HerreraWiarda[["pca"]]@stdev) * 100 # find standard deviation for each PC
cumu <- cumsum(pct) # find cumulative percentages for PCs
co1 <- which(cumu > 90 & pct < 5)[1] # find PC representing cumulative percent >90% and less than 5% associated with the single PC
co1 # list PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1 # find last PC where change in percent variation is more than 0.1%
co2 # list PC
pcs <- min(co1, co2) # find the minimum PC from the 2 methods used above
pcs # list PC
```

Plot our findings:

```{r, warning = FALSE, error = FALSE}
plot_df <- data.frame(pct = pct, # put PC values into dataframe for plotting
                      cumu = cumu, 
                      rank = 1:length(pct))
ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > pcs)) + # visualize PCs to use in elbow plot
  geom_text() + 
  geom_vline(xintercept = 90, color = "grey") + 
  geom_hline(yintercept = min(pct[pct > 5]), color = "grey") +
  theme_bw()
```

Store information for PCs to use in future computations:

```{r, warning = FALSE, error = FALSE}
PCdims <- 1:pcs # use the minimum PC from the quantitative method above to set the PCs for subsequent steps
length(PCdims) # how many significant PCs are there?
```

### Construct UMAP

Create a UMAP, using the calculated number of PCs from above:

```{r, warning = FALSE, error = FALSE, message = FALSE}
referenceSsPBMC_HerreraWiarda <- RunUMAP(referenceSsPBMC_HerreraWiarda, 
                               dims = PCdims, 
                               reduction = "pca", 
                               assay = "SCT") # create UMAP
```

Now check out plotting of some of the metadata groupings:

```{r, warning = FALSE, error = FALSE}
DimPlot(referenceSsPBMC_HerreraWiarda, 
        group.by = "SampleID") # plot by sample ID
DimPlot(referenceSsPBMC_HerreraWiarda, 
        group.by = "seurat_clusters") # plot by cell cluster numbers assigned in publication
```

### Identify and organize cells 

Change the order clusters are listed in under the seurat_clusters data slot:

```{r, warning = FALSE, error = FALSE}
Idents(referenceSsPBMC_HerreraWiarda) <- referenceSsPBMC_HerreraWiarda$seurat_clusters # change identities to cell annotations from publication
levels(referenceSsPBMC_HerreraWiarda) <- c('13', '19', '20', '25', '27', '30', '32', 
                                           '2', '7', '8', '10', '11', '15', '16', '23', '26', '33', '29',
                                           '0', '3', '4', '28', '9', '12', '14', '18', '22', 
                                           '5', '17', '1', '24', '31', '6', '21', '35')
referenceSsPBMC_HerreraWiarda$seurat_clusters <- Idents(referenceSsPBMC_HerreraWiarda) # apply new order to the identities in the Seurat object
```

Assign cell names from manuscript to the cluster numbers:

```{r, warning = FALSE, error = FALSE}
cellTypeList <- c(rep('Monocytes', 5), 'cDC', 'pDC', rep('B cells', 10), 'Antibody-secreting cells', rep('CD4pos ab T cells', 4), 
                  rep('CD8ABpos ab T cells', 5), rep('CD8Apos ab T/NK cells', 2), 'NK cells', rep('CD2pos gd T cells', 2), rep('CD2neg gd T cells', 2), 'Erythrocytes')
referenceSsPBMC_HerreraWiarda$cellType <-referenceSsPBMC_HerreraWiarda$seurat_clusters # make a new data slot based off cell groups
Idents(referenceSsPBMC_HerreraWiarda) <- referenceSsPBMC_HerreraWiarda$cellType
names(cellTypeList) <- levels(referenceSsPBMC_HerreraWiarda)
referenceSsPBMC_HerreraWiarda <- RenameIdents(referenceSsPBMC_HerreraWiarda, cellTypeList) # change dataset identity to cell types
referenceSsPBMC_HerreraWiarda$cellType <- Idents(referenceSsPBMC_HerreraWiarda)
```

Now plot these new cell classifications on our UMAP:

```{r, warning = FALSE, error = FALSE}
DimPlot(referenceSsPBMC_HerreraWiarda, # let's see where these cell types fall out in our 2-D visualization plot 
        group.by = "cellType", 
        label = TRUE)
```

Also assign cell lineages to cells:

```{r, warning = FALSE, error = FALSE}
Idents(referenceSsPBMC_HerreraWiarda) <- referenceSsPBMC_HerreraWiarda$cellType
cellLineageList <- c(rep('Myeloid lineage', 3), rep('B lineage', 2), rep('T/ILC lineage', 6), 'Non-immune')
referenceSsPBMC_HerreraWiarda$cellLineage <- referenceSsPBMC_HerreraWiarda$cellType # make a new data slot based off cell lineage groupings
Idents(referenceSsPBMC_HerreraWiarda) <- referenceSsPBMC_HerreraWiarda$cellLineage
names(cellLineageList) <- levels(referenceSsPBMC_HerreraWiarda)
referenceSsPBMC_HerreraWiarda <- RenameIdents(referenceSsPBMC_HerreraWiarda, cellLineageList) # change dataset identity to cell lineages
referenceSsPBMC_HerreraWiarda$cellLineage <- Idents(referenceSsPBMC_HerreraWiarda)
```

And plot cell lineage classifications onto the UMAP too:

```{r, warning = FALSE, error = FALSE}
DimPlot(referenceSsPBMC_HerreraWiarda, # let's see where these cell lineages fall out in our 2-D visualization plot again
        group.by = "cellLineage", 
        label = TRUE)
```

### Save the reference dataset as a Seurat object:

```{r, warning = FALSE, error = FALSE}
saveRDS(referenceSsPBMC_HerreraWiarda, '/Users/jayne.wiarda/Desktop/scRNAseq/scRNAseqIleumAtlas/Ileum/ReferenceData/HerreraWiarda_SsPBMC/reference_SsPBMC_HerreraWiarda.rds')
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

Integrated analysis of multimodal single-cell data. Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, Rahul Satija. bioRxiv 2020.10.12.335331; doi: https://doi.org/10.1101/2020.10.12.335331

Reference transcriptomes of porcine peripheral immune cells created through bulk and single-cell RNA sequencing
Juber Herrera-Uribe, Jayne E Wiarda, Sathesh K Sivasankaran, Lance Daharsh, Haibo Liu, Kristen A Byrne, Timothy Smith, Joan K Lunney, Crystal L Loving, Christopher K Tuggle
bioRxiv 2021.04.02.438107; doi: https://doi.org/10.1101/2021.04.02.438107

Elmentaite R, Ross ADB, Roberts K, James KR, Ortmann D, Gomes T, Nayak K, Tuck L, Pritchard S, Bayraktar OA, Heuschkel R, Vallier L, Teichmann SA, Zilbauer M. Single-Cell Sequencing of Developing Human Gut Reveals Transcriptional Links to Childhood Crohn's Disease. Dev Cell. 2020 Dec 21;55(6):771-783.e5. doi: 10.1016/j.devcel.2020.11.010.

H. Bengtsson, A Unifying Framework for Parallel and Distributed Processing in R using Futures, arXiv:2008.00553, 2020
 
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.
  
Heping Xu, Jiarui Ding, Caroline B.M. Porter, Antonia Wallrapp, Marcin Tabaka, Sai Ma, Shujie Fu, Xuanxuan Guo, Samantha J. Riesenfeld, Chienwen Su, Danielle Dionne, Lan T. Nguyen, Ariel Lefkovith, Orr Ashenberg, Patrick R. Burkett, Hai Ning Shi, Orit Rozenblatt-Rosen, Daniel B. Graham, Vijay K. Kuchroo, Aviv Regev, Ramnik J. Xavier. Transcriptional Atlas of Intestinal Immune Cells Reveals that Neuropeptide Î±-CGRP Modulates Group 2 Innate Lymphoid Cell Responses. Immunity, Volume 51, Issue 4, 2019, Pages 696-708.e9. doi: 10.1016/j.immuni.2019.09.004.